rtt_rosparam/include/rtt_rosparam/ros_param_data_source.hpp

```diff
@@ -0,0 +1,147 @@
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#ifndef __RTT_ROSPARAM__ROS_PARAM_DATA_SOURCE_HPP
+#define __RTT_ROSPARAM__ROS_PARAM_DATA_SOURCE_HPP
+
+#include "rtt/internal/DataSource.hpp"
+#include "rtt/internal/DataSourceTypeInfo.hpp"
+#include "rtt/internal/Reference.hpp"
+
+#include "ros/ros.h"
+#include "ros/param.h"
+#include "ros/names.h"
+
+#include <boost/function.hpp>
+#include <exception>
+
+using namespace RTT;
+
+namespace rtt_rosparam {
+
+/**
+ * A DataSource which checks the ROS parameter server and it
+ * returns the parameter value in its get() method.
+ * It sets the parameter value in its set() method.
+ * The constructor receives the full qualified property name attached
+ * to the DataSource.
+ * <br><b>Warning</b>
+ * The properties with a RosParamDataSource<T> are not real-time safe
+ * and therefore this DataSource shouldn't be used in the updateHook()
+ * of a real-time component.
+ */
+template<typename T>
+class RosParamDataSource
+    : public internal::AssignableDataSource<T>
+{
+  /**
+   * Stores the full-qualified property name of the ROS parameter
+   */
+  std::string mparam_name_;
+  mutable typename internal::DataSource<T>::value_t mcached_data_;
+
+public:
+  /**
+   * Use shared_ptr.
+   */
+  typedef boost::intrusive_ptr< RosParamDataSource<T> > shared_ptr;
+
+  ~RosParamDataSource()
+  {
+  }
+
+  RosParamDataSource(std::string param_name_id)
+      : mparam_name_(param_name_id),
+        mcached_data_()
+  {
+  }
+
+  typename internal::DataSource<T>::result_t get() const {
+    (void) evaluate();
+    return value();
+  }
+
+  bool evaluate() const
+  {
+    T return_value;
+    if (!ros::param::getCached(mparam_name_, return_value)) {
+      throw std::out_of_range("The parameter " + mparam_name_ + " cannot be fetched.");
+      return false;
+    }
+    mcached_data_ = return_value;
+    return true;
+  }
+
+  typename internal::DataSource<T>::result_t value() const
+  {
+    return mcached_data_;
+  }
+
+  // There is not referred element, no allocation exists for this data source and it is not an alias.
+  // But it needs to return something, because otherwise the .write() calls that require a rvalue()
+  // would fail. So we use an internal cache
+  typename internal::DataSource<T>::const_reference_t rvalue() const
+  {
+    return mcached_data_;
+  }
+
+  void set( typename internal::AssignableDataSource<T>::param_t t )
+  {
+    set() = t;
+    ros::param::set(mparam_name_, t);
+  }
+
+  // There is not referred element, no allocation exists for this data source and it is not an alias.
+  typename internal::AssignableDataSource<T>::reference_t set()
+  {
+    return mcached_data_;
+  }
```

Comment:
The class is missing an override of [AssignableDataSource<T>::updated()](https://github.com/orocos-toolchain/rtt/blob/600102e8be9c81905b20930e32d43b28244ab173/rtt/base/DataSourceBase.hpp#L148-L153). It is important to call that method after changing the cached value via `set()` without arguments. For example:
```cpp
RosParamDataSource<std::string> ds("~param");
functionThatTakesAReference(ds->set());
ds->updated();
```

```suggestion
  void set( typename internal::AssignableDataSource<T>::param_t t )
  {
    mcached_data_ = t;
    updated();
  }

  typename internal::AssignableDataSource<T>::reference_t set()
  {
    return mcached_data_;
  }

  void updated() {
    ros::param::set(mparam_name_, mcached_data_);
  } 
```